### DISCUSSION TITLE
prefetch && preload

### DISCUSSION THREAD

**Discussion by AndonMitev** [vercel/next.js#62937] [App Router] [answered]
Can someone explain the workings of prefetch and preload? Does prefetch only deal with the resource structure, while preload handles the actual data? Or does prefetch also take the data into account?

**Comment by icyJoseph** [comment]
Which prefetch/preload are we talking about here...

These?

- https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel/preload
- https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel/prefetch

Which you do with: https://react.dev/reference/react-dom/preload#noun-labs-1201738-(2) from the `react-dom` library.

Or are you talking about this bit: https://nextjs.org/docs/app/building-your-application/data-fetching/patterns#preloading-data

**Reply by AndonMitev** [reply]
Yes about Preload:

https://nextjs.org/docs/app/building-your-application/data-fetching/patterns#preloading-data

and for prefetch

https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#2-prefetching

Also i'm reading about swr

https://swr.vercel.app/docs/prefetching and they are mentioning that:

```
Together with techniques like [page prefetching(opens in a new tab)](https://nextjs.org/docs/api-reference/next/router#routerprefetch) in Next.js, you will be able to load both next page and data instantly.
```

My goal is to prefetch && preload everything and to bring instant user navigation

**Reply by icyJoseph** [reply, marked as answer]
Ok, so this onion has many layers!

- https://nextjs.org/docs/app/building-your-application/data-fetching/patterns#preloading-data

This one is talking about taking advantage of Next.js' fetch caching, which exists only sever side. 

So you'd just invoke the fetch process, and `void` the result, that is, not consume it. However, Next.js still completes the request, so that next time you do the same `fetch` call, you hit the Next.js fetch cache.

The hint is here:

> With this approach, you don't have to pass promises down as props. 

I've done that in some projects, where we don't use fetch, but direct DB promise handlers. I start the work in a server component, and pass the promise down to another server component as prop, and only there I await, and only this section is wrapped in a suspense boundary.

- https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#2-prefetching

This is a bit more of a Next.js optimization, which has existed since pages router.

Since some pages need JavaScript, and optionally, JSON, to be rendered, whenever there's a link within the viewport (plus minus some margin), Next.js goes and fetches the JavaScript, and potential JSON, for the page, so that when the user clicks the link, all that's left to do, is client side rendering.

In Pages Router, for pages with getServerSideProps, the JSON data would only be fetched on click.

And well,

- https://swr.vercel.app/docs/prefetching

The idea is pretty clear there, the preload link tag, gives the resource a high priority, so that it might as well be there by the time your app, client side, wants to access it, and SWR will be there to cache it into JavaScript memory, which is then used across your app to reduce re-rendering, since the same JavaScript memory is used. 

That's important because, even if the browser itself caches something, so that the next time you request it, is server from disk, you'd still allocate a new object to carry that data. 

So, yeah you could try to add preload tags to the HTML, that give higher priority to getting certain data, that could help your app be even snappier, I guess. I can't tell you which ones, but in general I think, `prefetch` from Next.js' Link does already quite a lot of heavy lifting for you. 

You might also want to read: https://react.dev/reference/react-dom/preload#noun-labs-1201738-(2)

**Reply by AndonMitev** [reply]
wow thank you so much for the explanation man, i would love to see application fully utilizing these patterns!
